Tools/Scripts/rewrite-compile-commands

```diff
@@ -0,0 +1,92 @@
+# LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR
+# A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT
+# OWNER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL,
+# SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT
+# LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE,
+# DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY
+# THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
+# (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+# OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+
+import argparse
+import json
+import os
+import re
+
+parser = argparse.ArgumentParser(
+    description='This tool takes the compile_commands.json output by CMake and expands any UnifiedSources in it to regular sources.'
+)
+parser.add_argument('input_file')
+parser.add_argument('output_file')
+parser.add_argument('source_dir')
+parser.add_argument('build_dir')
+
+args = parser.parse_args()
+source_dir = args.source_dir
+build_dir = args.build_dir
+
+generated_compile_commands = []
+
+with open(args.input_file) as f:
+    compile_commands = json.load(f)
+
+for entry in compile_commands:
+    entry_file = entry['file']
+    if not 'UnifiedSource' in entry_file:
+        generated_compile_commands.append(entry)
+        continue
+
+    # The paths we expect are in the style of:
+    # - WebKit/WebKitBuild/Release/DerivedSources/WebKit/unified-sources/UnifiedSource-54928a2b-29.cpp
+    # - WebKit/WebKitBuild/Release/WebCore/DerivedSources/unified-sources/UnifiedSource-42f7b70e-5.cpp
+    # And Windows too:
+    # - WebKit\WebKitBuild\Release\WebCore\DerivedSources\unified-sources\UnifiedSource-42f7b70e-5.cpp
+    # So we extract the parent folders to map them to the source folders.
+    parent_dir_1, parent_dir_2 = entry_file.rsplit(os.path.sep, maxsplit=4)[1:3]
+    search_directories = [
+        os.path.join(build_dir, parent_dir_1, parent_dir_2),
+    ]
+
+    if parent_dir_1 == 'DerivedSources':
+        search_directories.append(os.path.join(source_dir, 'Source', parent_dir_2))
+    else:
+        search_directories.append(os.path.join(source_dir, 'Source', parent_dir_1))
+
+    with open(entry_file) as f:
+        for line in f.readlines():
+            include_path = line[10:-2]  # Extract header from `#include "HEADER"\n`
+            for d in search_directories:
+                include_file = os.path.join(d, include_path)
+
+                if os.path.exists(include_file):
+                    generated_compile_commands.append({
+                        "directory": entry["directory"],
+                        "command": entry["command"].replace(entry_file, include_file),
+                        "file": include_file,
+                    })
+                    break
+            else:
+                print('Failed to find', include_path)
```

Comment:
This triggers for many files here, `*MessageReceiver.cpp` that don't exist and files from `WebCore/DerivedSources/` mostly.

Is this OK?

Reply:
Hmm in my local builds it found everything. What was your specific build config?

Reply:
Also was it ran *after* everything was built?

We are kinda in a weird situation where this script technically depends on every generated source, but I'm not sure I can tell CMake it should be the literal last target...

So maybe it runs out of order sometimes...

I didn't want to rely on checking for files existing but the rules of where a file ends up didn't have a clear pattern, but maybe some static mapping is possible.

Reply:
Just doing a clean GTK Debug build triggers the issue here, with sccache enabled, the script runs early on, I suppose before the MessageReceiver.cpp files have been generated.

Reply:
Actually those MessageReceiver.cpp files are for features we don't enable, such as GPUProcess, so it's expected that they shouldn't exist.

Reply:
> Actually those MessageReceiver.cpp files are for features we don't enable, such as GPUProcess, so it's expected that they shouldn't exist.

The file has to exist in some include path, it is compiled, otherwise the build would fail.